### PR TITLE
[Owner Review] Ported SynchronizeTclk example to python

### PR DESCRIPTION
**Pull request by ShantanuShinde**
### What does this Pull Request accomplish?

Ported the **SynchronizeTClk** C example to python using grpc APIs. This example makes use of both Fgen and TClk APIs

### Why should this Pull Request be merged?

This example can be used to test NI-FGEN and NI-TClk APIs and also acts as a demonstration

### What testing has been done?

The example was tested on PXI-5412 Fgen hardware

### CODE REVIEW

File: examples/nitclk/synchronize-tclk.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+# Demonstrates how to configure multiple NI Arbitrary Waveform Generators to generate synchronized waveforms with TClk
+# 
+# The gRPC API is built from the C API. NI-FGEN documentation is installed with the driver at:
+# C:\Program Files (x86)\IVI Foundation\IVI\Drivers\niFgen\documentation\English\SigGenHelp.chm
+#
+# The gRPC API is built from the C API. NI-TClk documentation is installed with the driver at:
```
Comment (doshirohan) [inline examples/nitclk/synchronize-tclk.py]: You don't need to repeat `The gRPC API is built from the C API.` two times. You can remove this from TClk documentation here.

File: examples/nitclk/synchronize-tclk.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+#   > python synchronize-tclk.py <server_address> <port_number>
+# If they are not passed in as command line arguments, then by default the server address will be "localhost:31763"
+
+import grpc
+import sys
+import nifgen_pb2 as nifgen_types
+import nifgen_pb2_grpc as grpc_nifgen
+import nitclk_pb2 as nitclk_types
+import nitclk_pb2_grpc as grpc_nitclk
+import session_pb2 as session_types
```
Comment (doshirohan) [inline examples/nitclk/synchronize-tclk.py]: I don't see either of session imports being used anywhere. Do you need these?

File: examples/nitclk/synchronize-tclk.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+# To run this example, install "NI-FGEN Driver" on the server machine.
+# Link : https://www.ni.com/en-us/support/downloads/drivers/download.ni-fgen.html
+#
+# For instructions on how to use protoc to generate gRPC client interfaces, see our "Creating a gRPC Client" wiki page.
+# Link: https://github.com/ni/grpc-device/wiki/Creating-a-gRPC-Client
+#
+# Running from command line:
+#
+# Server machine's IP address and port number can be passed as separate command line arguments.
+#   > python synchronize-tclk.py <server_address> <port_number>
```
Comment (harsha-bhushan) [inline examples/nitclk/synchronize-tclk.py]: have similar comment to other tclk example https://github.com/doshirohan/grpc-device/blob/features/tclk-digitalpattern/examples/nitclk/multi-device-configured-acquisition-tclk.py

File: examples/nitclk/synchronize-tclk.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import nitclk_pb2_grpc as grpc_nitclk
+import session_pb2 as session_types
+import session_pb2_grpc as grpc_session
+import matplotlib.pyplot as plt
+
+server_address = "localhost"
+server_port = "31763"
+
+# Take resource name . This example doesn't work on simulation, so resource name is required.
+resources = input("Enter comma separated resource names (PXI1Slot2,PXI1Slot3): ").split(",")
```
Comment (harsha-bhushan) [inline examples/nitclk/synchronize-tclk.py]: take in resource as arguments similar to other tclk example. what would happen if only 1 resource is passed?

File: examples/nitclk/synchronize-tclk.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+def ThrowOnError (service, vi, error_code):
+    if service == nifgen_service:
+        error_message_request = nifgen_types.ErrorMessageRequest(
+            vi = vi,
+            error_code = error_code
+        )
+        error_message_response = nifgen_service.ErrorMessage(error_message_request)
+        raise Exception (error_message_request.error_message)
+    else:
+        error_message_request = nitclk_types.GetExtendedErrorInfoRequest()
```
Comment (doshirohan) [inline examples/nitclk/synchronize-tclk.py]: Name these variables as `error_info_request` and `error_info_response` to match with request and response objects you are creating.

File: examples/nitclk/synchronize-tclk.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        error_message_response = nitclk_service.GetExtendedErrorInfo(error_message_request)           
+        raise Exception (error_message_response.error_string)
+
+try:
+    # list of sessions
+    sessions = []
+    i = 0
+    for resource in resources:
+        # Initalize NI-FGEN session
+        init_with_channels_resp = nifgen_service.InitWithOptions(nifgen_types.InitWithOptionsRequest(
```
Comment (doshirohan) [inline examples/nitclk/synchronize-tclk.py]: Change variable name to match API you are calling.

File: examples/nitclk/synchronize-tclk.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+# If NI-FGEN API throws an exception, print the error message  
+except grpc.RpcError as rpc_error:
+    error_message = rpc_error.details()
+    if rpc_error.code() == grpc.StatusCode.UNAVAILABLE :
+        error_message = f"Failed to connect to server on {server_address}"
+    elif rpc_error.code() == grpc.StatusCode.UNIMPLEMENTED:
+        error_message = "The operation is not implemented or is not supported/enabled in this service"
+    print(f"{error_message}") 
+finally:
+    if 'vi' in vars() and vi.id != 0:
```
Comment (doshirohan) [inline examples/nitclk/synchronize-tclk.py]: This needs to be updated to close all sessions we have opened. You will need to iterate over `sessions` and whichever of them have non-zero `id`, call `Close` on them.

**Review by doshirohan — CHANGES_REQUESTED** [review]

File: examples/nitclk/synchronize-tclk.py (hunk 4 above)
Comment (doshirohan) [reply examples/nitclk/synchronize-tclk.py]: Difference between this and tclk-scope example is, FGEN can't be simulated. So users will always have to pass in resource name (and hence server address and port number) if we make it a parameter. If we make `resources` as first parameter, then parameter list won't match other examples.

If user inputs only one resource name, `split` will create array of one element.

Comment (doshirohan) [reply examples/nitclk/synchronize-tclk.py]: With that, I think the comment and input message here needs to be updated. If Harsha is ok with taking user input, you should change above lines to

```
# Read FGEN resources to be configured
 resources = input("Enter comma separated FGEN resource names (PXI1Slot2, PXI1Slot3): ").split(",").strip()
```

The `strip` call would remove any leading/trailing whitespaces.

File: examples/nitclk/synchronize-tclk.py (hunk 7 above)
Comment (ShantanuShinde) [reply examples/nitclk/synchronize-tclk.py]: changed that

File: examples/nitclk/synchronize-tclk.py (hunk 6 above)
Comment (ShantanuShinde) [reply examples/nitclk/synchronize-tclk.py]: fixed

File: examples/nitclk/synchronize-tclk.py (hunk 2 above)
Comment (ShantanuShinde) [reply examples/nitclk/synchronize-tclk.py]: removed

File: examples/nitclk/synchronize-tclk.py (hunk 4 above)
Comment (ShantanuShinde) [reply examples/nitclk/synchronize-tclk.py]: changed the input part

File: examples/nitclk/synchronize-tclk.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+import grpc
+import sys
+import nifgen_pb2 as nifgen_types
+import nifgen_pb2_grpc as grpc_nifgen
+import nitclk_pb2 as nitclk_types
+import nitclk_pb2_grpc as grpc_nitclk
+import matplotlib.pyplot as plt
+
+server_address = "10.164.64.53"
```
Comment (doshirohan) [inline examples/nitclk/synchronize-tclk.py]: This should be `localhost`

**Review by doshirohan — APPROVED** [review]

File: examples/nitclk/synchronize-tclk.py (hunk 3 above)
Comment (ShantanuShinde) [reply examples/nitclk/synchronize-tclk.py]: changed this

File: examples/nitclk/synchronize-tclk.py (hunk 4 above)
Comment (ShantanuShinde) [reply examples/nitclk/synchronize-tclk.py]: changed this to be same as other examples

File: examples/nitclk/synchronize-tclk.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+# Link : https://www.ni.com/en-us/support/downloads/drivers/download.ni-fgen.html
+#
+# For instructions on how to use protoc to generate gRPC client interfaces, see our "Creating a gRPC Client" wiki page.
+# Link: https://github.com/ni/grpc-device/wiki/Creating-a-gRPC-Client
+#
+# Running from command line:
+#
+# Server machine's IP address, port number, and multiple comma separated resource names can be passed as separate command line arguments.
+#   > python synchronize-tclk.py <server_address> <port_number> <resource_name>
+# This example doesn't work on simulation, so these arguments are mandatory, not optional
```
Comment (harsha-bhushan) [inline examples/nitclk/synchronize-tclk.py]: Change the comment to - This example is not supported in simulation mode, hence these arguments are mandatory

File: examples/nitclk/synchronize-tclk.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+# To run this example, install "NI-FGEN Driver" on the server machine.
+# Link : https://www.ni.com/en-us/support/downloads/drivers/download.ni-fgen.html
+#
+# For instructions on how to use protoc to generate gRPC client interfaces, see our "Creating a gRPC Client" wiki page.
+# Link: https://github.com/ni/grpc-device/wiki/Creating-a-gRPC-Client
+#
+# Running from command line:
+#
+# Server machine's IP address, port number, and multiple comma separated resource names can be passed as separate command line arguments.
+#   > python synchronize-tclk.py <server_address> <port_number> <resource_name>
```
Comment (harsha-bhushan) [inline examples/nitclk/synchronize-tclk.py]: <resource_name1,resource_name2>

File: examples/nitclk/synchronize-tclk.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+# Create waveform data
+waveform_data = []
+for i in range(waveform_size//2):
+    waveform_data.append(0.0)
+for i in range(waveform_size//2, waveform_size):
+    waveform_data.append(1.0)
+
+# Read in cmd args
+if len(sys.argv) < 4:
+    print("This example doesn't work on simulation, please provide server address, server port and resource name as follows:")
```
Comment (harsha-bhushan) [inline examples/nitclk/synchronize-tclk.py]: This example is not supported in simulation mode. Please provide...

File: examples/nitclk/synchronize-tclk.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    sys.exit("python synchronize-tclk.py <server_address> <port_number> <resource_name>")
```
Comment (harsha-bhushan) [inline examples/nitclk/synchronize-tclk.py]: <resource_name1,resource_name2>

**Review by harsha-bhushan — APPROVED** [review]

**Review by doshirohan — APPROVED** [review]

File: examples/nitclk/synchronize-tclk.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+    print(f"Generating square wave with sample rate {sample_rate} on {resources}...")
+    print("Close the graph or press Ctrl+C to stop generation")
+
+    try:
+        # Plot waveform
+        fig = plt.gcf()
+        fig.canvas.manager.set_window_title('Sample Waveform')
+        plt.plot(waveform_data)
+        plt.suptitle("Close the window to stop generation", fontsize=10)
```
Comment (jonlugoatx) [inline examples/nitclk/synchronize-tclk.py]: Space out the `=` operator.

File: examples/nitclk/synchronize-tclk.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        pass
+
+# If NI-FGEN API throws an exception, print the error message  
+except grpc.RpcError as rpc_error:
+    error_message = rpc_error.details()
+    if rpc_error.code() == grpc.StatusCode.UNAVAILABLE :
+        error_message = f"Failed to connect to server on {server_address}"
+    elif rpc_error.code() == grpc.StatusCode.UNIMPLEMENTED:
+        error_message = "The operation is not implemented or is not supported/enabled in this service"
+    print(f"{error_message}")
```
Comment (jonlugoatx) [inline examples/nitclk/synchronize-tclk.py]: No need to format here: `print(error_message)`

File: examples/nitclk/synchronize-tclk.py [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+import nitclk_pb2 as nitclk_types
+import nitclk_pb2_grpc as grpc_nitclk
+import matplotlib.pyplot as plt
+
+# parameters
+sample_rate = 20000000.0
+waveform_size = 16
+# Create waveform data
+waveform_data = []
+for i in range(waveform_size//2):
```
Comment (jonlugoatx) [inline examples/nitclk/synchronize-tclk.py]: Space out the `//` operator.

**Review by jonlugoatx — APPROVED** [review]

**Review by kyledsullivan — APPROVED** [review]

**Pull request merged**